treeopt/treeOpt.py

```diff
@@ -0,0 +1,277 @@
+        :param limits: Array containing the highest and lowest limits for each
+        dimension
+        :type limits: Numpy-Array
+        :return: Nothing
+        :rtype: None
+
+        """
+
+        self.limits = limits
+
+    def set_simulate_method(self, method):
+        """
+        Sets the method, which should be used to start the Simulation
+        :param method: A keyword representing the method
+        the simulation String
+        :type method: Python function
+        :return: Nothing
+        :rtype: None
+
+        """
+        self.simKeyword = method
+        if self.simKeyword == "benchmarking":
+            self.simulateMethod = simulate.simulate_benchmark_function
+        elif self.simKeyword == "python":
+            self.simulateMethod = simulate.simulate_benchmark_function
+        elif self.simKeyword == "extern":
+            self.simulateMethod = simulate.simulateExternalProgramm
+        else:
+            print("Error")
+
+    def set_python_problem(self, problem):
+        """
+        Stores a Function, that represents a benchmarking problem
+        :param problem: Python function, which returns points of a responce
+        surface
+        :type problem: Python function
+        :return: Nothing
+        :rtype: None
+
+        """
+
+        self.benchmarkingProblem = problem
+
+    def set_sm_method(self, method):
+        """
+        Sets the Method that is used to approximate the system responce in the
+        design space
+        :param method: one of the functions in modules/metamodell.py
+        :type method: python function
+        :return: Nothing
+        :rtype: None
+
+        """
+
+        self.smMethod = method
+
+    def set_accuracy_criterion(self, method):
+        """
+        Sets the method that is used to define the termination condition of the
+        optimization workflow
+        :param method: one of the functions in modules/accuracy
+        :type method: Python Function
+        :return: Nothing
+        :rtype: None
+
+        """
+
+        self.accuracyCriterion = method
+
+    # Starts one simulation run
+    def simulate(self, x):
+        """
+        Function that starts a Simulation and returns the resonce of the system
+        :param x: Numpy array representing a set of parameters to be written on
+        the simulation Input file
+        :type x: Numpy array
+        :return: Numpy array containing the system responce
+        :rtype: Numpy array
+
+        """
+        try:
+            if self.simulateMethod == simulate.simulate_benchmark_function:
+                return self.simulateMethod(self.benchmarkingProblem, x)
+            if self.simulateMethod == simulate.simulateExternalProgramm:
+                return self.simulateMethod(
+                    self.inputFile,
+                    self.outputFile,
+                    self.simFile,
+                    self.program,
+                    x,
+                )
+        except TypeError:
```

Comment:
You want to throw an error if `self.simulateMethod` is neither `simulate.simulate_benchmark_function` nor `simulate.simulateExternalProgram`, right? Then you shouldn't use a try-catch block, but you should throw a `TypeError` here. 

This error can then be catched by whoever uses the `simulate` function to prevent a crash.

Reply:
Ah, I just found my comment from before. So in addition to taking care of the situation where self.simulateMethod` is something we didn't expect, we should handle the cases where the call of `simulateMethod` fails. 

If it is a python function, we can try to catch whatever exceptions are thrown from within the python function, if any exceptions are thrown. Since we do not know the exception type, we should do something like the accepted answer here: https://stackoverflow.com/questions/30442236/how-to-prevent-too-broad-exception-in-this-case

Now if we are calling an external program using `subprocess`, we should call it with `check==True` to make sure a `CalledProcessError` is raised. This exception can than be caught here (or elsewhere, just make sure your program never crashes. Exit with clear and understandable error messages/exceptions).